pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/ConcurrentOpenLongPairRangeSet.java

```diff
@@ -0,0 +1,474 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.common.util.collections;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.NavigableMap;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.StampedLock;
+
+import org.apache.pulsar.common.util.collections.LongPairRangeSet.LongPairConsumer;
+import org.apache.pulsar.common.util.collections.LongPairRangeSet.RangeProcessor;
+
+import com.google.common.collect.BoundType;
+import com.google.common.collect.Range;
+
+/**
+ * A Concurrent set comprising zero or more ranges of type {@link LongPair}. This can be alternative of
+ * {@link com.google.common.collect.RangeSet} and can be used if {@code range} type is {@link LongPair} </br>
+ * 
+ * <pre>
+ *  
+ * Usage:
+ * a. This can be used if one doesn't want to create object for every new inserted {@code range}
+ * b. It creates {@link BitSet} for every unique first-key of the range. 
+ * So, this rangeSet is not suitable for large number of unique keys.
+ * </pre>
+ * 
+ *
+ */
+public class ConcurrentOpenLongPairRangeSet<T extends Comparable<T>> implements LongPairRangeSet<T> {
+
+    protected final NavigableMap<Long, BitSet> rangeBitSetMap = new ConcurrentSkipListMap<>();
+    private boolean threadSafe = true;
+    private final int bitSetSize;
+    private final LongPairConsumer<T> consumer;
+
+    // caching place-holder for cpu-optimization to avoid calculating ranges again
+    private volatile int cachedSize = 0;
+    private volatile String cachedToString = "[]";
+    private volatile boolean updatedAfterCached = true;
+
+    public ConcurrentOpenLongPairRangeSet(LongPairConsumer<T> consumer) {
```

Comment:
Can you clarify what is the purpose of the "consumer"? Since it's not typical to have a consumer on a set.

Reply:
`ConcurrentOpenLongPairRangeSet<T>` stores `T (PositionImpl)` into bits. However, many of the apis require converter to create `T (PositionImpl)` from the stored bit. eg: `Range<T> firstRange(), void forEach(RangeProcessor<T> action), Range<T> span(), Collection<Range<T>> asRanges()` so, instead adding `Converter-Consumer` in signature of all apis, I have added as part of constructor.
Also, `com.google.common.collect.RangeSet` doesn't use this converter-consumer because it directly stores `T(PositionImpl)` and retrieves it from the stored data-structure. So, we have  [LongPairRangeSet<T>](https://github.com/apache/pulsar/blob/01e873fe14936c321c587002f7d8f40b648d69b4/pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/LongPairRangeSet.java) interface which mostly follows `com.google.common.collect.RangeSet` api-definition and i don't want to change them by adding `converter-consumer` in each of those methods.